src/solvers/dg/2d/dg.jl

```diff
@@ -1933,144 +1910,90 @@ function calc_mortar_flux!(surface_flux_values, dg::Dg2D, mortar_type::Val{:l2},
-        end
-      else # large_sides[m] == 2 -> large element on right side
-        @views large_surface_flux_values = (dg.l2mortar_reverse_upper *
-                                        (fstar_upper[v, :] + noncons_diamond_upper[v,:]) +
-                                            dg.l2mortar_reverse_lower *
-                                        (fstar_lower[v, :] + noncons_diamond_lower[v,:]))
-        if dg.l2mortars.orientations[m] == 1
-          # L2 mortars in x-direction
-          surface_flux_values[v, :, 1, large_element_id] .= large_surface_flux_values
-        else
-          # L2 mortars in y-direction
-          surface_flux_values[v, :, 3, large_element_id] .= large_surface_flux_values
-        end
-      end
-    end
+    @. fstar_upper += noncons_diamond_upper
+    @. fstar_lower += noncons_diamond_lower
+    copy_and_project_mortar_fluxes!(surface_flux_values, dg, mortar_type, m,
+                                    fstar_upper, fstar_lower)
   end
 end
 
-# Calculate and store fluxes across L2 mortars
-function calc_mortar_flux!(surface_flux_values, dg::Dg2D, mortar_type::Val{:l2},
-                           nonconservative_terms::Val{false}, mortars, cache)
-  @unpack neighbor_ids, u_lower, u_upper, orientations = mortars
-  @unpack fstar_upper_threaded, fstar_lower_threaded = cache
-
-  Threads.@threads for m in 1:dg.n_l2mortars
-    large_element_id = dg.l2mortars.neighbor_ids[3, m]
-    upper_element_id = dg.l2mortars.neighbor_ids[2, m]
-    lower_element_id = dg.l2mortars.neighbor_ids[1, m]
-
-    # Choose thread-specific pre-allocated container
-    fstar_upper = fstar_upper_threaded[Threads.threadid()]
-    fstar_lower = fstar_lower_threaded[Threads.threadid()]
-
-    # Calculate fluxes
-    calc_fstar!(fstar_upper, u_upper, m, orientations, dg)
-    calc_fstar!(fstar_lower, u_lower, m, orientations, dg)
-
-    # Copy flux small to small
-    if dg.l2mortars.large_sides[m] == 1 # -> small elements on right side
-      if dg.l2mortars.orientations[m] == 1
-        # L2 mortars in x-direction
-        surface_flux_values[:, :, 1, upper_element_id] .= fstar_upper
-        surface_flux_values[:, :, 1, lower_element_id] .= fstar_lower
-      else
-        # L2 mortars in y-direction
-        surface_flux_values[:, :, 3, upper_element_id] .= fstar_upper
-        surface_flux_values[:, :, 3, lower_element_id] .= fstar_lower
-      end
-    else # large_sides[m] == 2 -> small elements on left side
-      if dg.l2mortars.orientations[m] == 1
-        # L2 mortars in x-direction
-        surface_flux_values[:, :, 2, upper_element_id] .= fstar_upper
-        surface_flux_values[:, :, 2, lower_element_id] .= fstar_lower
-      else
-        # L2 mortars in y-direction
-        surface_flux_values[:, :, 4, upper_element_id] .= fstar_upper
-        surface_flux_values[:, :, 4, lower_element_id] .= fstar_lower
-      end
+@inline function copy_and_project_mortar_fluxes!(surface_flux_values, dg::Dg2D, mortar_type::Val{:l2}, m,
```

Comment:
This method needs a good docstring, since a lot of "infrastructure" work is hidden in here

Reply:
:+1: